crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+use std::cmp;
+use std::fmt::{self, Display};
+use std::io::Write;
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+use virtio_queue::DescriptorChain;
+use vm_memory::{Bytes, GuestMemory, GuestMemoryError};
+
+/// Virtio console chain processing abstraction.
+///
+/// This module offers an abstraction for working with the descriptor chains of the receive and
+/// transmit queues.
+///  * The transmit queue is used to send data from the driver to the device and it's buffers are
+///    read-only for the device.
+///  * The receive queue is used to send data from the device to the driver and it's buffers are
+///    write-only for the device.
+///
+/// To implement this abstraction the [`Console`](struct.Console.html)  structure is provided which
+/// consists of the `input_buffer` member and exposes several methods.
+/// * The [`input_buffer`](struct.Console.html#field.input_buffer) is a buffer used to store data
+///   before sending it to the driver. The following methods are provided to interact with it:
+///   - [`enqueue_data`](struct.Console.html#method.enqueue_data) which is used to add data to the
+///     end of the buffer.
+///   - [`is_input_buffer_empty`](struct.Console.html#method.is_input_buffer_empty) which is used
+///     assess if there is any data waiting to be sent.
+///
+/// The methods for processing a descriptor chain are the following:
+/// *  [`process_transmitq_chain`](struct.Console.html#method.process_transmitq_chain) which is
+///    used to read the data from a descriptor chain of the transmit queue.
+/// * [`process_receiveq_chain`](struct.Console.html#method.process_receiveq_chain) which is used
+///   to send the data to the driver by writing it into the buffers of a receive queue descriptor
+///   chain.
+
+/// Console device errors.
+#[derive(Debug)]
+pub enum Error {
+    /// Failed to write slice to guest memory.
+    WriteToGuestFailed(GuestMemoryError),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            WriteToGuestFailed(ref err) => {
+                write!(f, "console failed to write slice to guest memory: {}", err)
+            }
+        }
+    }
+}
+
+/// Console struct that implements the abstraction of virtio console descriptor chain handling.
+pub struct Console {
```

Comment:
Can you make sure to derive all the useful traits? Examples and explanations are available here: https://rust-lang.github.io/api-guidelines/interoperability.html

Reply:
I added `Default`, `Clone` and `Debug`. Is there a need for anything else?